Libraries/LibGemini/Line.cpp

```diff
@@ -0,0 +1,150 @@
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <AK/StringBuilder.h>
+#include <LibGemini/Document.h>
+
+namespace Gemini {
+
+String Text::render_to_html() const
+{
+    StringBuilder builder;
+    builder.append(m_text);
+    builder.append("<br>\n");
+    return builder.build();
+}
+
+Text::~Text()
+{
+}
+
+String Heading::render_to_html() const
+{
+    StringBuilder builder;
+    size_t level = 0;
+    while (m_text.length() > level && m_text[level] == '#')
+        ++level;
+
+    builder.appendf("<h%d>", level);
+    builder.append(m_text.substring_view(level, m_text.length() - level));
+    builder.appendf("</h%d>", level);
+    return builder.build();
+}
+Heading::~Heading()
+{
+}
+
+String UnorderedList::render_to_html() const
+{
+    // 1.3.5.4.2 "Advanced clients can take the space of the bullet symbol into account"
+    //           FIXME: _where_ is the space?
+    StringBuilder builder;
+    builder.append("<li>");
+    builder.append(m_text.substring_view(1, m_text.length() - 1));
+    builder.append("</li>");
+    return builder.build();
+}
+UnorderedList::~UnorderedList()
+{
+}
+
+String Control::render_to_html() const
+{
+    switch (m_kind) {
+    case Kind::PreformattedEnd:
+        return "</pre>";
+    case Kind::PreformattedStart:
+        return "<pre>";
+    case Kind::UnorderedListStart:
+        return "<ul>";
+    case Kind::UnorderedListEnd:
+        return "</ul>";
+    }
+
+    dbg() << "Unknown control kind _" << m_kind << "_";
+    ASSERT_NOT_REACHED();
+    return "";
+}
+Control::~Control()
+{
+}
+
+Link::Link(String text, const Document& document)
+    : Line(move(text))
+{
+    size_t index = 2;
+    while (index < m_text.length() && (m_text[index] == ' ' || m_text[index] == '\t'))
+        ++index;
+    auto url_string = m_text.substring_view(index, m_text.length() - index);
+    auto space_offset = url_string.find_first_of(" \t");
+    String url = url_string;
+    if (space_offset.has_value()) {
+        url = url_string.substring_view(0, space_offset.value());
+        auto offset = space_offset.value();
+        while (offset < url_string.length() && (url_string[offset] == ' ' || url_string[offset] == '\t'))
+            ++offset;
+        m_name = url_string.substring_view(offset, url_string.length() - offset);
+    }
+    m_url = URL(url);
+    if (!m_url.is_valid()) {
+        // relative URL
```

Comment:
I wouldn't rely on `URL::is_valid()` for telling relative vs absolute URLs...

Reply:
any alternatives?

Reply:
Maybe add a `URL::is_absolute()`, IDK :D

Reply:
That's quite an annoying thing to do, since it would effectively duplicate a lot of stuff that `URL::complete` is doing, e.g. `//blah.blah/blah` is absolute

Reply:
It seems `complete_url` already handles this, so I'm just gonna drop this bit off